go-client/client/app_stub.go

```diff
@@ -0,0 +1,88 @@
+package client
+
+import (
+	"context"
+	"fmt"
+
+	apppb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/application"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type appStub struct {
+	client *Client
+
+	stub apppb2.ApplicationServiceClient
+
+	cancelFunc context.CancelFunc
+}
+
+func newAppStub(client *Client) appStub {
+	return appStub{client: client, stub: apppb2.NewApplicationServiceClient(client.grpcChannel)}
+}
+
+type changeHandler func(update *apppb2.FieldsChangeUpdate)
+
+// Note that the provided context is saved and used to continue reading the stream
+func (as *appStub) listFields(ctx context.Context, handler changeHandler) error {
+	as.cancelListLoop()
+
+	ctx, cancel := context.WithCancel(as.client.withToken(ctx))
+
+	req := apppb2.ListFieldsRequest{}
+	fieldStream, err := as.stub.ListFields(ctx, &req)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	resp, err := fieldStream.Recv()
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	as.client.handleFieldChanges(resp)
+
+	as.cancelFunc = cancel
+
+	lister := fieldLister{stream: fieldStream, handler: handler}
+	go lister.listLoop()
+
+	return nil
+}
+
+func (as *appStub) isListing() bool {
+	return as.cancelFunc != nil
+}
+
+func (as *appStub) cancelListLoop() {
+	if as.cancelFunc != nil {
+		as.cancelFunc()
+		as.cancelFunc = nil
+	}
+}
+
+func (as *appStub) Close() {
+	as.cancelListLoop()
+}
+
+type fieldLister struct {
+	stream  apppb2.ApplicationService_ListFieldsClient
+	handler changeHandler
+}
+
+func (fl *fieldLister) listLoop() {
+	for {
+		resp, err := fl.stream.Recv()
+		if err != nil {
+			if status, ok := status.FromError(err); ok && status.Code() == codes.Canceled {
+				break
+			}
+			fmt.Println("failed to list fields: ", err)
+			return
```

Comment:
Calling FetchTables with FetchOnce is always guaranteed to synchronously return a single error. Calling FetchTables will FetchRepeating's first error (if it happens) will always be synchronous too, but every error after that will be async. So there are some possibilities:
1. Split the function into a `FetchTablesOnce()` and a `FetchTablesRepeating()`, so that one can return a simple error and one can return an error channel.
2. Keep the single function call, and have it return an error channel that might just immediately close when using FetchOnce.